### PR TITLE
Add PerSAM

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the PerSAM model.

Question: when you do:

```
from transformers import PerSamModel

model = PerSamModel.from_pretrained("facebook/sam-vit-huge")
```

you get this warning:

```
You are using a model of type sam to instantiate a model of type persam. This is not supported for all configurations of models and can yield errors.
```

was wondering whether we could suppress this warning. PerSAM uses the exact same weights as the original SAM model, just modifies the forward pass with 2 additional arguments. Currently the model_type is set to "persam" in `PerSamConfig`.

### CODE REVIEW

File: src/transformers/models/sam/modeling_sam.py [hunk 1]
```diff
@@ -576,8 +576,12 @@ def __init__(self, config: SamPromptEncoderConfig):
...
         self.conv2 = nn.Conv2d(self.mask_input_channels, config.mask_input_channels, kernel_size=2, stride=2)
         self.conv3 = nn.Conv2d(config.mask_input_channels, config.hidden_size, kernel_size=1)
-        self.layer_norm1 = SamLayerNorm(self.mask_input_channels, config.layer_norm_eps)
-        self.layer_norm2 = SamLayerNorm(self.mask_input_channels * 4, config.layer_norm_eps)
+        self.layer_norm1 = SamLayerNorm(
+            self.mask_input_channels, eps=config.layer_norm_eps, data_format="channels_first"
+        )
+        self.layer_norm2 = SamLayerNorm(
+            self.mask_input_channels * 4, eps=config.layer_norm_eps, data_format="channels_first"
+        )
```
Comment (sgugger) [inline src/transformers/models/sam/modeling_sam.py:584]: @younesbelkada can you confirm if this a mistake in the SAM implementation?

**Review by sgugger — COMMENTED** [review]
Thanks for adding this. If the modification works with all checkpoints for SAM then it passes the test for which we don't require a new model. So in this instance it's fine to add a new config argument to SAM to activate persam mode and adapt the forward, which will also get rid of your warning.

File: src/transformers/models/sam/modeling_sam.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/sam/modeling_sam.py:584]: Yes I had to include this fix to make `input_masks` work. I noticed that `input_masks` is not tested in `tests/models/test_modeling_sam.py` (only `input_points`, `input_labels` and `input_boxes` are).

Might be good to add a test for this

**Comment by NielsRogge** [comment]
Ok, will close this PR in favor of modifying `modeling_sam.py`.

File: src/transformers/models/sam/modeling_sam.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/models/sam/modeling_sam.py:584]: Sounds good to me! Indeed we didn't tested input masks!
